File: cognitive/src/main/java/mine/ui/RecordFragment.java
package mine.ui;

import android.graphics.Typeface;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.content.ContextCompat;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;

import com.example.cognitive.R;
import com.example.common.persistense.risk.DailyRiskEntity;
import com.example.common.record.ui.RecordDetailBottomSheet;
import com.github.mikephil.charting.charts.LineChart;
import com.github.mikephil.charting.components.XAxis;
import com.github.mikephil.charting.components.YAxis;
import com.github.mikephil.charting.data.Entry;
import com.github.mikephil.charting.data.LineData;
import com.github.mikephil.charting.data.LineDataSet;
import com.github.mikephil.charting.formatter.ValueFormatter;

import java.time.LocalDate;
import java.util.ArrayList;
import java.util.List;
import java.util.TreeSet;

import mine.vm.RecordViewModel;

public class RecordFragment extends Fragment {

    private static final String TAG = "RecordFragment";

    private TextView week;
    private TextView half;
    private LineChart lineChart;
    private RecyclerView record;
    private SwipeRefreshLayout swipeRefresh;
    private RecordRVAdapter adapter;
    private List<DailyRiskEntity> riskDataList = new ArrayList<>();
    private int selectedDays = 15;

    private RecordViewModel viewModel;

    public RecordFragment() {
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_record, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        viewModel = new ViewModelProvider(this).get(RecordViewModel.class);

        bindView(view);
        bindClickListener();
        initLineChart();
        initRecyclerView();
        observeViewModel();
    }

    private void observeViewModel() {
        viewModel.getAllRiskData().observe(getViewLifecycleOwner(), list -> {
            if (list.isEmpty()) {
                return;
            }
            riskDataList = new ArrayList<>(list);
            adapter.setList(riskDataList);
            updateLineChartData(riskDataList);
            swipeRefresh.setRefreshing(false);
        });
    }

    private void bindView(View view) {
        week = view.findViewById(R.id.week);
        half = view.findViewById(R.id.half);
        record = view.findViewById(R.id.record);
        lineChart = view.findViewById(R.id.lineChart);
        swipeRefresh = view.findViewById(R.id.swipeRefresh);
    }

    private void bindClickListener() {
        week.setOnClickListener(v -> {
            selectedDays = 7;
            updateButtonAppearance();
            viewModel.queryRecordsByDays(7);
        });
        half.setOnClickListener(v -> {
            selectedDays = 15;
            updateButtonAppearance();
            viewModel.queryRecordsByDays(15);
        });
        updateButtonAppearance();

        swipeRefresh.setOnRefreshListener(() -> viewModel.queryRecordsByDays(selectedDays));
    }

    private void updateButtonAppearance() {
        if (selectedDays == 7) {
            week.setBackgroundTintList(ContextCompat.getColorStateList(requireContext(), R.color.blue));
            half.setBackgroundTintList(ContextCompat.getColorStateList(requireContext(), R.color.deep_grey));
        } else {
            week.setBackgroundTintList(ContextCompat.getColorStateList(requireContext(), R.color.deep_grey));
            half.setBackgroundTintList(ContextCompat.getColorStateList(requireContext(), R.color.blue));
        }
    }

    private void initLineChart() {
        lineChart.getDescription().setEnabled(false);
        lineChart.getLegend().setEnabled(false);
        lineChart.setTouchEnabled(true);
        lineChart.setDragEnabled(true);
        lineChart.setScaleEnabled(false);
        lineChart.setDrawGridBackground(false);

        Typeface tf = Typeface.createFromAsset(requireContext().getAssets(), "fonts/harmonyos_sans_regular.ttf");

        XAxis xAxis = lineChart.getXAxis();
        xAxis.setPosition(XAxis.XAxisPosition.BOTTOM);
        xAxis.setDrawGridLines(false);
        xAxis.setTypeface(tf);
        xAxis.setGranularity(1f);

        YAxis yAxis = lineChart.getAxisLeft();
        yAxis.setDrawGridLines(true);
        yAxis.setTypeface(tf);
        yAxis.setGranularity(0.1f);
        lineChart.getAxisRight().setEnabled(false);

        updateLineChartData(riskDataList);
    }

    private void initRecyclerView() {
        adapter = new RecordRVAdapter(new ArrayList<>());
        record.setLayoutManager(new LinearLayoutManager(getContext()));
        record.setAdapter(adapter);

        com.example.common.util.ItemSpacingDecoration itemSpacingDecoration =
                new com.example.common.util.ItemSpacingDecoration(getContext(), 3, 16, false);
        record.addItemDecoration(itemSpacingDecoration);

        adapter.setOnRecordClickListener((position, riskEntity) -> {
            FragmentManager fm = getParentFragmentManager();
            if (fm.findFragmentByTag("RecordDetailBottomSheet") != null) {
                return;
            }
            LocalDate date = riskEntity.getDate();
            RecordDetailBottomSheet sheet = RecordDetailBottomSheet.newInstance(date);
            sheet.show(getParentFragmentManager(), "RecordDetailBottomSheet");
        });

        viewModel.queryRecordsByDays(selectedDays);
    }

    private void updateLineChartData(List<DailyRiskEntity> list) {
        List<Entry> entries = new ArrayList<>();
        if (list != null && !list.isEmpty()) {
            for (int i = 1; i < list.size(); i++) {
                DailyRiskEntity entity = list.get(i);
                entries.add(new Entry(i, (float) entity.getRiskScore()));
            }
        }
        LineDataSet dataSet = new LineDataSet(entries, "");
        dataSet.setColor(ContextCompat.getColor(requireContext(), R.color.blue));
        dataSet.setLineWidth(1f);
        dataSet.setDrawCircles(false);
        dataSet.setDrawValues(false);

        lineChart.setData(new LineData(dataSet));

        TreeSet<Float> yLabels = new TreeSet<>();
        for (Entry entry : entries) {
            float val = entry.getY();
            float floor = (float) (Math.floor(val * 10) / 10.0);
            float ceil = (float) (Math.ceil(val * 10) / 10.0);
            yLabels.add(floor);
            yLabels.add(ceil);
        }
        YAxis yAxis = lineChart.getAxisLeft();
        if (!yLabels.isEmpty()) {
            yAxis.setAxisMinimum(yLabels.first());
            yAxis.setAxisMaximum(yLabels.last());
            yAxis.setValueFormatter(new ValueFormatter() {
                @Override
                public String getFormattedValue(float value) {
                    return String.format("%.1f", value);
                }
            });
        } else {
            yAxis.setAxisMinimum(0f);
            yAxis.setAxisMaximum(1f);
        }
        yAxis.setGranularityEnabled(true);

        lineChart.invalidate();
    }
}

File: bridge/src/main/java/com/example/bridge/record/ui/RecordFragment.kt
package com.example.bridge.record.ui

import android.graphics.Typeface
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import android.widget.Toast
import androidx.core.content.ContextCompat
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentManager
import androidx.lifecycle.lifecycleScope
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
import com.example.bridge.R
import com.example.common.login.GuestStateHolder
import com.example.common.login.remote.LoginStatusManager
import com.example.common.persistense.AppDatabase
import com.example.common.persistense.risk.DailyRiskEntity
import com.example.common.persistense.risk.RiskRepository
import com.example.common.repository.network.BindApiService
import com.example.common.repository.network.RetrofitClient
import com.example.common.record.ui.RecordDetailBottomSheet
import com.example.common.util.ItemSpacingDecoration
import com.github.mikephil.charting.charts.LineChart
import com.github.mikephil.charting.components.XAxis
import com.github.mikephil.charting.components.YAxis
import com.github.mikephil.charting.data.Entry
import com.github.mikephil.charting.data.LineData
import com.github.mikephil.charting.data.LineDataSet
import com.github.mikephil.charting.formatter.ValueFormatter
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.time.LocalDate
import java.util.TreeSet
import kotlin.math.ceil
import kotlin.math.floor

class RecordFragment : Fragment() {

    private var week: TextView? = null
    private var half: TextView? = null
    private var lineChart: LineChart? = null
    private var record: RecyclerView? = null
    private var swipeRefresh: SwipeRefreshLayout? = null
    private var adapter: RecordRVAdapter? = null
    private var riskDataList: MutableList<DailyRiskEntity> = ArrayList()
    private var selectedDays = 15

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        return inflater.inflate(R.layout.fragment_record, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        Log.d("RecordFragment", "onViewCreated")

        try {
            bindView(view)
            bindClickListener()
            initLineChart()
            initRecyclerView()
        } catch (e: Exception) {
            Log.e("RecordFragment", "Error in onViewCreated", e)
            throw e
        }
    }

    private fun bindView(view: View) {
        week = view.findViewById(R.id.week)
        half = view.findViewById(R.id.half)
        record = view.findViewById(R.id.record)
        lineChart = view.findViewById(R.id.lineChart)
        swipeRefresh = view.findViewById(R.id.swipeRefresh)
    }

    private fun bindClickListener() {
        week?.setOnClickListener {
            selectedDays = 7
            updateButtonAppearance()
            loadRecordDataWithNetworkFirst()
        }
        half?.setOnClickListener {
            selectedDays = 15
            updateButtonAppearance()
            loadRecordDataWithNetworkFirst()
        }
        updateButtonAppearance()

        swipeRefresh?.setOnRefreshListener {
            refreshFromNetwork()
        }
    }

    private fun refreshFromNetwork() {
        if (GuestStateHolder.isGuest()) {
            Log.d("RecordFragment", "游客模式，直接刷新本地数据")
            loadRiskDataFromDatabase()
            swipeRefresh?.isRefreshing = false
            return
        }

        val eldername = LoginStatusManager.getLoggedInUserId(requireContext())
        if (eldername.isEmpty()) {
            Log.w("RecordFragment", "未登录，无法获取绑定设备数据")
            swipeRefresh?.isRefreshing = false
            return
        }

        viewLifecycleOwner.lifecycleScope.launch(Dispatchers.IO) {
            try {
                val apiService = RetrofitClient.createService(BindApiService::class.java)
                val networkData = apiService.getAllDailyRisk(eldername)

                if (networkData.isNotEmpty()) {
                    Log.d("RecordFragment", "从网络获取到 ${networkData.size} 条风险数据")
                    val database = AppDatabase.getDatabase(requireContext())
                    database.dailyRiskDao().insertAll(networkData)
                    Log.d("RecordFragment", "已更新本地数据库")

                    withContext(Dispatchers.Main) {
                        Toast.makeText(context, "同步成功", Toast.LENGTH_SHORT).show()
                        loadRiskDataFromDatabase()
                        swipeRefresh?.isRefreshing = false
                    }
                } else {
                    Log.d("RecordFragment", "网络数据为空")
                    withContext(Dispatchers.Main) {
                        // TODO：暂时注释
                        // Toast.makeText(context, "暂无数据", Toast.LENGTH_SHORT).show()
                        swipeRefresh?.isRefreshing = false
                    }
                }
            } catch (e: Exception) {
                if (e is CancellationException) throw e
                Log.e("RecordFragment", "从网络获取数据失败", e)
                withContext(Dispatchers.Main) {
                    swipeRefresh?.isRefreshing = false
                    Toast.makeText(context, "同步失败", Toast.LENGTH_SHORT).show()
                }
            }
        }
    }

    private fun loadRecordDataWithNetworkFirst() {
        viewLifecycleOwner.lifecycleScope.launch {
            swipeRefresh?.isRefreshing = true
            try {
                val (data, networkSuccess) = withContext(Dispatchers.IO) {
                    loadRecordData()
                }
                if (view == null) return@launch
                updateUI(data)
                if (networkSuccess) {
                    Toast.makeText(requireContext(), "同步成功", Toast.LENGTH_SHORT).show()
                } else {
                    // TODO：暂时注释
                    // Toast.makeText(requireContext(), "暂无数据", Toast.LENGTH_SHORT).show()
                }
            } catch (e: Exception) {
                if (e is CancellationException) throw e
                Log.e("RecordFragment", "加载数据失败: ${e.message}", e)
                Toast.makeText(requireContext(), "刷新失败", Toast.LENGTH_SHORT).show()
            } finally {
                swipeRefresh?.isRefreshing = false
            }
        }
    }

    private suspend fun loadRecordData(): Pair<List<DailyRiskEntity>, Boolean> = withContext(Dispatchers.IO) {
        val eldername = LoginStatusManager.getLoggedInUserId(requireContext())
        if (eldername.isEmpty()) {
            return@withContext Pair(emptyList(), false)
        }

        // 游客模式直接读本地，不走网络
        if (GuestStateHolder.isGuest()) {
            Log.d("RecordFragment", "游客模式，直接读取本地数据")
            val database = AppDatabase.getDatabase(requireContext())
            val fromDate: LocalDate = LocalDate.now().minusDays((selectedDays - 1).toLong())
            val list = RiskRepository.getFromBlocking(database.dailyRiskDao(), fromDate)
            val result: MutableList<DailyRiskEntity> = ArrayList(list)
            result.reverse()
            return@withContext Pair(result, false)
        }

        // 网络接口拉取和存库
        var networkSuccess = false
        try {
            val apiService = RetrofitClient.createService(BindApiService::class.java)
            val networkData = apiService.getAllDailyRisk(eldername)
            if (networkData.isNotEmpty()) {
                val database = AppDatabase.getDatabase(requireContext())
                database.dailyRiskDao().insertAll(networkData)
                Log.d("RecordFragment", "网络数据已存库: ${networkData.size} 条")
                networkSuccess = true
            }
        } catch (e: Exception) {
            if (e is CancellationException) throw e
            Log.e("RecordFragment", "网络拉取失败，读取本地缓存: ${e.message}")
        }

        // 读库
        val database = AppDatabase.getDatabase(requireContext())
        val fromDate: LocalDate = LocalDate.now().minusDays((selectedDays - 1).toLong())
        val list = RiskRepository.getFromBlocking(database.dailyRiskDao(), fromDate)
        val result: MutableList<DailyRiskEntity> = ArrayList(list)
        result.reverse()
        Pair(result, networkSuccess)
    }

    private fun updateUI(list: List<DailyRiskEntity>) {
        riskDataList = list.toMutableList()
        adapter?.setList(list)
        updateLineChartData(list)
    }

    private fun updateButtonAppearance() {
        if (selectedDays == 7) {
            week?.backgroundTintList = ContextCompat.getColorStateList(requireContext(), R.color.blue)
            half?.backgroundTintList = ContextCompat.getColorStateList(requireContext(), R.color.deep_grey)
        } else {
            week?.backgroundTintList = ContextCompat.getColorStateList(requireContext(), R.color.deep_grey)
            half?.backgroundTintList = ContextCompat.getColorStateList(requireContext(), R.color.blue)
        }
    }

    private fun initLineChart() {
        lineChart?.description?.isEnabled = false
        lineChart?.legend?.isEnabled = false
        lineChart?.setTouchEnabled(true)
        lineChart?.isDragEnabled = true
        lineChart?.setScaleEnabled(false)
        lineChart?.setDrawGridBackground(false)

        val tf = Typeface.createFromAsset(requireContext().assets, "fonts/harmonyos_sans_regular.ttf")

        val xAxis: XAxis = lineChart!!.xAxis
        xAxis.position = XAxis.XAxisPosition.BOTTOM
        xAxis.setDrawGridLines(false)
        xAxis.typeface = tf
        xAxis.granularity = 1f

        val yAxis: YAxis = lineChart!!.axisLeft
        yAxis.setDrawGridLines(true)
        yAxis.typeface = tf
        yAxis.granularity = 0.1f
        lineChart?.axisRight?.isEnabled = false

        updateLineChartData(riskDataList)
    }

    private fun initRecyclerView() {
        adapter = RecordRVAdapter(ArrayList())
        record?.layoutManager = LinearLayoutManager(context)
        record?.adapter = adapter

        val itemSpacingDecoration = ItemSpacingDecoration(requireContext(), 3, 16, false)
        record?.addItemDecoration(itemSpacingDecoration)

        adapter?.setOnRecordClickListener { _, riskEntity ->
            val fm: FragmentManager = parentFragmentManager
            if (fm.findFragmentByTag("RecordDetailBottomSheet") != null) {
                return@setOnRecordClickListener
            }

            val date: LocalDate = riskEntity.date
            val sheet = RecordDetailBottomSheet.newInstance(date)
            sheet.show(parentFragmentManager, "RecordDetailBottomSheet")
        }

        loadRecordDataWithNetworkFirst()
    }

    private fun loadRiskDataFromDatabase() {
        viewLifecycleOwner.lifecycleScope.launch(Dispatchers.IO) {
            try {
                val database = AppDatabase.getDatabase(requireContext())
                val fromDate: LocalDate = LocalDate.now().minusDays((selectedDays - 1).toLong())
                val list = RiskRepository.getFromBlocking(database.dailyRiskDao(), fromDate)

                val result: MutableList<DailyRiskEntity> = ArrayList(list)
                result.reverse()

                withContext(Dispatchers.Main) {
                    riskDataList = result
                    adapter?.setList(result)
                    updateLineChartData(result)
                }
            } catch (e: Exception) {
                if (e is CancellationException) throw e
                e.printStackTrace()
            }
        }
    }

    private fun updateLineChartData(list: List<DailyRiskEntity>?) {
        val entries: MutableList<Entry> = ArrayList()
        var x = 0
        if (list != null && !list.isEmpty()) {
            for (i in list.size - 1 downTo  0) {
                val entity = list[i]
                entries.add(Entry((++x).toFloat(), entity.riskScore.toFloat()))
            }
        }
        val dataSet = LineDataSet(entries, "")
        dataSet.color = ContextCompat.getColor(requireContext(), R.color.blue)
        dataSet.lineWidth = 1f
        dataSet.setDrawCircles(false)
        dataSet.setDrawValues(false)

        lineChart?.data = LineData(dataSet)

        // Update Y axis range and labels
        val yLabels = TreeSet<Float>()
        for (entry in entries) {
            val value = entry.y
            val floor = (floor((value * 10).toDouble()) / 10.0).toFloat()
            val ceil = (ceil((value * 10).toDouble()) / 10.0).toFloat()
            yLabels.add(floor)
            yLabels.add(ceil)
        }
        val yAxis: YAxis? = lineChart?.axisLeft
        if (yLabels.isNotEmpty()) {
            yAxis?.axisMinimum = yLabels.first()
            yAxis?.axisMaximum = yLabels.last()
            yAxis?.valueFormatter = object : ValueFormatter() {
                override fun getFormattedValue(value: Float): String {
                    return String.format("%.1f", value)
                }
            }
        } else {
            yAxis?.axisMinimum = 0f
            yAxis?.axisMaximum = 1f
        }
        yAxis?.isGranularityEnabled = true

        lineChart?.invalidate()
    }
}
